### ISSUE TITLE
OMP_NUM_THREADS=1 much faster than unset OMP_NUM_THREADS

### ISSUE THREAD

**Issue by ChessPoker** [deeplearning4j/libnd4j#287]
OMP_NUM_THREADS=1 seems to be required on both my machines ("bart" and "loco"). I have built from master on each. I have also built javacpp from master as well. I'm using Intel MKL on both.

On "bart"
With OMP_NUM_THREADS unset, an epoch takes 124 seconds. With OMP_NUM_THREADS=1 an epoch takes 23 seconds. That is about a 5 to 6 times difference.

On "loco" 
With OMP_NUM_THREADS unset, an epoch takes 665 seconds. With OMP_NUM_THREADS=1 an epoch takes 27 seconds. That is about a 25 times difference.

loco VisualVM hotspots with unset OMP_NUM_THREADS:
![hotspots](https://cloud.githubusercontent.com/assets/8595703/18263031/933c09f6-73bb-11e6-9030-4b6b3b310317.png)
To get decent performance I have to use ParallelWrapper with 10 to 14 workers and OMP_NUM_THREADS=1. Even then I'm still only using about half the cpu capacity. In this case, however, I'm able to complete an epoch in under 4 seconds.

"loco" is local Linux Mint 17.3
CPU:  2 x Intel Xeon E5-2687Wv3, 3.1 GHz (10-Core, HT, 25MB Cache, 160W) 22nm
RAM:  64GB (8 x 8GB DDR4-2133 ECC Registered 1R 1.2V DIMMs) Operating at 2133 MT/s Max
"bart" is remote Centos 7.2
CPU:  2 x Intel Xeon E5-2643v3, 3.4 GHz (6-Core, HT, 20MB Cache, 135W) 22nm
RAM:  64GB (8 x 8GB DDR4-2133 ECC Registered 1R 1.2V DIMMs) Operating at 2133 MT/s Max

With OMP_NUM_THREADS=1 the performance differential is about what you'd expect given the two chips clock frequency (3.4 vs 3.1). However, there is a lot of difference between the two machines with OMP_NUM_THREADS unset. I assume on loco unset is equivalent to OMP_NUM_THREADS=40 and on bart unset is equivalent to OMP_NUM_THREADS=24. I think I will follow up with OMP_NUM_THREADS=24 on loco.
.

**Comment by ChessPoker** [comment]
On loco and OMP_NUM_THREADS=24 the time for the first few epochs are:
epoch 0: 467 seconds
epoch 1:523 seconds
epoch 2:444 seconds
epoch 3:260 seconds
Not sure what is going on. VisualVm output below:
![hotspots](https://cloud.githubusercontent.com/assets/8595703/18264119/414f1072-73c3-11e6-9ec9-dd2014ba412c.png)

**Comment by raver119** [comment]
Are you checking 0.5.0 or current master?

**Comment by raver119** [comment]
I mean: issue is known one, was addressed already, and i'm kinda surprised to see pairwise being topped that high.

**Comment by ChessPoker** [comment]
I'm using 0.5.1-SNAPSHOT from yesterday. Less than 24 hours ago I git cloned everything.

**Comment by raver119** [comment]
Ok. Reduce being topped is somewhat expected. But pairwise... It shouldn't be using that much time.

Show me call tree screenshot please.

**Comment by ChessPoker** [comment]
Call Tree:
![calltree](https://cloud.githubusercontent.com/assets/8595703/18264779/1abbd446-73c7-11e6-8fe2-6491f7b03a9f.png)

**Comment by ChessPoker** [comment]
Some results from experiments:

OMP_NUM_THREADS=4
OMP_PROC_BIND=true
Epoch times (secs):
36.6, 26.7, 25.3, 25.1
108 ms per execTransfromFloat()
106 ms per execPairwiseTransformFloat()


OMP_NUM_THREADS=1
OMP_PROC_BIND=true
34.2, 27.9, 24.1, 23.3
104 ms per execTransformFloat()
107 ms per execPairwiseTransformFloat()

OMP_NUM_THREADS=4
OMP_PROC_BIND=true
GOMP_CPU_AFFINITY=0 2 4 6
39.3, 28.4, 25.8, 24.5

OMP_NUM_THREADS=4
GOMP_CPU_AFFINITY=0 2 4 6
34.8, 27.7, 24.1, 23.9

OMP_NUM_THREADS=4
37.4, 27.8, 23.7, 24.0

OMP_NUM_THREADS=1
38.3, 29.9, 25.7, 25.9
108 ms per T
106 ms per PWT

For all of these tests I see no significant difference in cpu utilization -- around 110%. This is strange.
I expected OMP_NUM_THREADS=4 to use more CPU than OMP_NUM_THREADS=1.

There is no evidence that any of these runs above are actually different. There is no evidence supporting that OMP_NUM_THREADS=4 invokes more threads than OMP_NUM_THREADS=1. We don't see it in cpu utilization, we don't see it in the composition of HotSpots and we don't see it in the ms/invocation of some key native calls. 

The time differences above are quite likely just noise. The experimental setup is just my typical training setup. It's not a benchmarking setup. The DataSet order changes between runs and there is a validation pass about 10% the size of the training set.

If there were parallel computation attempts when OMP_NUM_THREADS=4 that were not attempted when OMP_NUM_THREADS=1, would we not see some difference somewhere in some of these metrics?
Typical HotSpot view (all runs had about same CPU time percentages for top hotspots):
![hotspots](https://cloud.githubusercontent.com/assets/8595703/18292427/f28df8c0-7441-11e6-9ec9-5c0315da081f.png)

**Comment by ChessPoker** [comment]
@raver119 requested more experimental data:
All trials are 4 epochs and there is no validation set pass. It's all training.
==1==
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
OMP_NUM_THREADS=1
~ 100% CPU
120 seconds.

==2==
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
OMP_NUM_THREADS=4
~200% CPU
121 seconds.

==3==
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
~800% CPU
138 seconds.

==4==
MKL_THREADING_LAYER=GNU
~2000% CPU
313 seconds.

==5==
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=CLOSE
~800% CPU
142 seconds.

==6==
MKL_THREADING_LAYER=GNU
GOMP_CPU_AFFINITY=0 2 4 6
OMP_NUM_THREADS=4
~200% CPU
122 seconds.

I've set aside this benchmarking code so that I can rerun it in future. I can add VisualVM screenshots if needed.

**Comment by treo** [comment]
Could you try replacing the GNU OpenMP Runtime with Intels OpenMP?
You can download it from https://www.openmprtl.org/sites/default/files/libomp_20160808_oss.tgz and build it with the instructions they provide. There is no need to run the install step.

Then create a new empty folder (e.g. `/home/username/lib`), and link your newly built Intel OpenMP Runtime `(src/libiomp5.so)` to `libgomp.so.1` in that empty folder. And then start your benchmark code with the additional environment variables (esp. keep the MKL_THREADING_LAYER=GNU )
```
LD_LIBRARY_PATH=/home/username/lib:$LD_LIBRARY_PATH
OMP_DISPLAY_ENV=VERBOSE
```

And add the JVM parameter: `-Dorg.nd4j.nativeblas.pathsfirst=true`

You should see a option dump when openmp initializes. You should see only one of them, and there should be some variables that start with `KMP_`, that way you know you are using the correct open mp runtime.

**Comment by ChessPoker** [comment]
==7==
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=1
~150% CPU
117 seconds.

==8==
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=4
~275% CPU
119 seconds.

**Comment by treo** [comment]
@ChessPoker is that with intel's openmp? or just more results for raver?

**Comment by ChessPoker** [comment]
More results for @raver119 
I think we will try your idea to see what happens. Looks like it'll be easy to revert if need be.

**Comment by treo** [comment]
Yes, reverting back to default can be done by simply removing/reverting the LD_LIBRARY_PATH environment variable.

**Comment by ChessPoker** [comment]
Just so we are all clear. To date, these env variables were also present:
LD_LIBRARY_PATH=/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin/
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libgomp.so.1

**Comment by treo** [comment]
Ok, then you should already have a recent enough libiomp5.so somewhere in /opt/intel, so you can skip the compilation step, just create an empty directory and link it there as explained above. And then remove the LD_PRELOAD env variable. 

When you start the benchmark with OMP_DISPLAY_ENV=VERBOSE, you should see only the config dump for a single OpenMP Runtime, if you are seeing two of them, something is wrong.

**Comment by ChessPoker** [comment]
There's two: 
/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin_mic/libiomp5.so
/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin/libiomp5.so

**Comment by treo** [comment]
Use the second.

**Comment by ChessPoker** [comment]
$pwd
/home/mib/dl4jstack/altlib
$ls -lh
libgomp.so.1 -> /opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin/libiomp5.so

ENV VARS:
LD_LIBRARY_PATH=/home/mib/dl4jstack/altlib:/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=4
OMP_DISPLAY_ENV=VERBOSE

JVM: <clipped> -Dorg.nd4j.nativeblas.pathsfirst=true

**Comment by ChessPoker** [comment]
KMP_ in the option dump.
Only one option dump.

**Comment by treo** [comment]
that looks right, so how is the timing?

**Comment by ChessPoker** [comment]
(the '.1' refers to same as '8' above but using libgomp.so.1 -> /opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin/libiomp5.so)
== 8.1 ==
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=4
~450%
107 seconds

8.0 was 119s so it's looking better.

**Comment by raver119** [comment]
what about the case w/o OMP_NUM_THREADS set?

**Comment by ChessPoker** [comment]
== 7.1 ==
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=1
~150%
103 seconds.

**Comment by ChessPoker** [comment]
== 3.1 ==
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
.... cpu started at ~400% but in the last epoch it **pegged to 3900%**
* = * = * = * = * = * = * = * = * = ** EPOCH 0 ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 0 COMPLETE (32.7 seconds) ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 1 ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 1 COMPLETE (31.5 seconds) ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 2 ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 2 COMPLETE (31.6 seconds) ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 3 ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
* = * = * = * = * = * = * = * = * = ** EPOCH 3 COMPLETE (35.6 seconds) ** = * = * = * = * = * = * = * = * = * = * = * = * = * =
Elapsed time = 131

**Comment by treo** [comment]
Can you add a Hotspot Screenshot?

**Comment by ChessPoker** [comment]
![hotspots](https://cloud.githubusercontent.com/assets/8595703/18309425/3b49c8fc-74ae-11e6-9ec0-4bb79f52a0af.png)

**Comment by ChessPoker** [comment]
![calltree0](https://cloud.githubusercontent.com/assets/8595703/18309464/6d6d886e-74ae-11e6-80ea-bd7571f30d51.png)

**Comment by ChessPoker** [comment]
want to use libsvml.so
LD_LIBRARY_PATH=/home/mib/dl4jstack/altlib:/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin:/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin
== 3.2 ==
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
103 seconds.

3.1 was 131 seconds. 3.0 was 138 seconds.

But...
CPU pegs to 3900% in the final epoch.
![hotspots](https://cloud.githubusercontent.com/assets/8595703/18310768/489fb26c-74b5-11e6-805e-5e08c8cdd8d6.png)

![calltree0](https://cloud.githubusercontent.com/assets/8595703/18310795/6d358b60-74b5-11e6-8e05-8f6d25b252fb.png)

**Comment by ChessPoker** [comment]
== 2.2 ==
LD_LIBRARY_PATH=/home/mib/dl4jstack/altlib:/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin:/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
OMP_NUM_THREADS=4
95 seconds.

== 1.2 ==
LD_LIBRARY_PATH=/home/mib/dl4jstack/altlib:/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin:/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin
MKL_THREADING_LAYER=GNU
OMP_PROC_BIND=true
OMP_NUM_THREADS=1
71 seconds

7.2
LD_LIBRARY_PATH=/home/mib/dl4jstack/altlib:/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin:/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=1
80 seconds.

**Comment by treo** [comment]
Ok, so it looks like there is still some kind of thread overhead bottleneck there.

**Comment by ChessPoker** [comment]
Figured out some further improvements. First baseline test:
MKL_THREADING_LAYER=GNU (always present for all runs; omitting below for readability)
OMP_NUM_THREADS=1
80 seconds. Same result as above.

Next, make sure worker thread is on same processor (whole socket not a core) as master thread:
OMP_NUM_THREADS=1
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
69 seconds.

Next, try more threads:
OMP_NUM_THREADS=4
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
68 seconds.

Try even more threads (there are 10 cores per socket):
OMP_NUM_THREADS=8
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
66 seconds.

**Progress!**

Fine tune number of threads:
OMP_NUM_THREADS=6
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
67 seconds.

OMP_NUM_THREADS=9
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
68 seconds.

OMP_NUM_THREADS=7
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
63 seconds.
**Best result is 63 seconds**, down from 80 seconds (or 71 seconds depending on which you compare too).

There's a big boost simply by instructing workers not to wonder off to a distant processor (socket), even when num_threads=1. This I assume means even in single threaded mode a worker thread is spawned (just one). Seems stupid, but maybe there's a reason for that. Or maybe I assumed wrong.

There's a further boost by increasing the num_threads. So parallelism is working, finally. The reason it wasn't working before is, I assume, workers were spread far and wide, with some ending up on the other processor and having to load data from main memory. When workers reside on the same processor, cache hits are probably helping.

**Comment by raver119** [comment]
What you have if you don't specify OMP_NUM_THREADS?

**Comment by ChessPoker** [comment]
Baseline test (I'll run a baseline consisting of a known fast config anytime significant time has passed):
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
OMP_NUM_THREADS=7
65 seconds. (63 seconds last time)
Not sure why that is 65 seconds now. I've changed nothing. I ran test twice to confirm (first run may be initializing disk caches or something).

OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
156 seconds.
I think because, when not set, OMP_NUM_THREADS [defaults](https://gcc.gnu.org/onlinedocs/libgomp/OMP_005fNUM_005fTHREADS.html#OMP_005fNUM_005fTHREADS) to number of CPUs.  Here I think CPU means core. So it's running with OMP_NUM_THREADS = 20; or 10 if it looks at only one socket. No matter whether it's 10 or 20, because workers are confined to the same socket as the master, workers will only get access to 10 cores.

**Comment by raver119** [comment]
That's easy to check. Run with OMP_NUM_THREADS=10, and see if final time will be 156 seconds. But i'm afraid it'll be equal to =20 :(

**Comment by raver119** [comment]
Just to be clear: it's our code set's OMP_NUM_THREADS, and if env var is unset, we set it to number of total cores available. That should be changed then.

**Comment by ChessPoker** [comment]
Baseline:
OMP_PLACES=sockets
OMP_PROC_BIND=master, master,master
OMP_NUM_THREADS=7
65 seconds. (same as last time)

OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
OMP_NUM_THREADS=10
64 seconds. 

OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
OMP_NUM_THREADS=20
84 seconds.

OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
OMP_NUM_THREADS=40
150 seconds.

Looks like unsetting OMP_NUM_THREADS is similar to OMP_NUM_THREADS=40. Perhaps the code is setting OMP_NUM_THREADS=40, not 20. 40 is the number of logical threads (2 x 10 cores; each core has HT). 

The GNU libgomp default is: if ["undefined one thread per CPU is used"](https://gcc.gnu.org/onlinedocs/libgomp/OMP_005fNUM_005fTHREADS.html#OMP_005fNUM_005fTHREADS) which I think means that if no intervention was done in our code, OMP_NUM_THREADS would be set to 20 by default on my system.

**Comment by raver119** [comment]
So, i've merged r119_omp_reduce to master now, and will investigate how we can better set default values to OMP_NUM_THREADS together with affinity enforcement

**Comment by raver119** [comment]
@ChessPoker poke me in gitter when you'll be online, i have things to test :)

**Comment by raver119** [comment]
Additionally better default num_threads value selection was added to current master.

**Comment by ChessPoker** [comment]
branch r119_affinity tests:
LD_LIBRARY_PATH=/home/mib/dl4jstack/altlib:/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin:/opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin
MKL_THREADING_LAYER=GNU
OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
OMP_NUM_THREADS=7
OMP_DISPLAY_ENV=verbose
66 seconds (1 second slower than last time)
=====
OMP_NUM_THREADS=1
81 seconds (1 second slower than last time too). 
======
OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
OMP_NUM_THREADS=10
66 seconds (2 seconds slower than last time).
======
No Environment Variables set except for:
LD_LIBRARY_PATH, MKL_THREADING_LAYER and OMP_DISPLAY_ENV.
135 seconds.
=====
Same, but add:
OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
137 seconds.

**Comment by ChessPoker** [comment]
1) Built and installed OpenBLAS from source. Installed to /opt/OpenBLAS. Exported PATH=/opt/OpenBLAS/lib:$PATH.
2) Built libnd4j. Build output is [here](https://gist.github.com/ChessPoker/f81c3723e6bf352e993004a7e1ccb2b4).

= 1 = 
MKL_THREADING_LAYER=GNU
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libgomp.so.1
2000% CPU
306 seconds.

= 2= 
added 
OMP_NUM_THREADS=10
350% CPU
132 seconds.

= 3 = 
added 
OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
350% CPU
120 seconds.

= 4 = 
MKL_THREADING_LAYER=GNU
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libgomp.so.1
OMP_NUM_THREADS=20
OMP_PLACES=sockets
OMP_PROC_BIND=master,master,master
900% CPU
228 seconds.
= 5 = 
changed OMP_NUM_THREADS=1
150% CPU
94 seconds.
 = 6 = 
changed OMP_NUM_THREADS=4
275% CPU
96 seconds.
 = 7 = 
changed OMP_NUM_THREADS=7
300% CPU
121 seconds.

= 8 = 
**Changed to MKL** by adding
LD_LIBRARY_PATH=/opt/intel/compilers_and_libraries_2016.3.210/linux/mkl/lib/intel64_lin and creating symbolic links: libopenblas.so -> libmkl_rt.so, libopenblas.so.0 -> libmkl_rt.so, libblas.so.3 -> libmkl_rt.so.
Rebuilt ND4J with above LD_LIBRARY_PATH
121 seconds. 
= 9 = 
**Switched to Intel's implementation of OMP** by adding /home/mib/dl4jstack/altGOMP/ to LD_LIBRARY_PATH and adding -Dorg.nd4j.nativeblas.pathsfirst=true to JVM params.
93 seconds.

= 10 =
**use libsvml.so and edited CMakeLists.txt** as per @treo suggestion from above
added  /opt/intel/compilers_and_libraries_2016.3.210/linux/compiler/lib/intel64_lin
to LD_LIBRARY_PATH
65 seconds.

**Comment by raver119** [comment]
```
= 2= 
added 
OMP_NUM_THREADS=10
***350% CPU***
132 seconds.
```
That makes no sense for me.

**Comment by ChessPoker** [comment]
I believe the last comment had two OpenMP runtimes running for most of the tests. I think the MKL tests are valid, but the OpenBLAS stuff I think are no good.
[libnd4j build output](https://gist.github.com/ChessPoker/8bb122aea6a98823b2db1f1752f74777)

Using OpenBLAS.
=1=
MKL_THREADING_LAYER=GNU
2600% CPU
369 seconds.
=2=
Added OMP_NUM_THREADS=10
1200% CPU
136 seconds.
=3=
Added OMP_PLACES=sockets and OMP_PROC_BIND=master,master,master
1250% CPU
134 seconds.

=4= 
MKL_THREADING_LAYER=GNU
OMP_DISPLAY_ENV=verbose
OMP_NUM_THREADS=7
850% CPU
139 seconds
=5=
MKL_THREADING_LAYER=GNU
OMP_NUM_THREADS=1
105 seconds.
=6=
Switched to Intel's implementation of OMP by adding /home/mib/dl4jstack/altGOMP/ to LD_LIBRARY_PATH and adding -Dorg.nd4j.nativeblas.pathsfirst=true to JVM params.
98 seconds.
=7=
Recompiled libnd4j same as last except add "-march=native" compiler option
85 seconds.
=8=
change OMP_NUM_THREADS=10 (formerly 1)
1900% CPU
100 seconds.
=9=
Added OMP_PLACES=sockets and OMP_PROC_BIND=master,master,master
1900% CPU
95 seconds
=10=
Reduced OMP_NUM_THREADS=5
975% CPU
77 seconds.

Why is CPU % roughly double 100 x OMP_NUM_THREADS? The console prints "o.n.n.NativeOps - Number of threads used for linear algebra 5" **twice**. Are 5 threads running on each socket?

Anyway, seems that we're getting close to the same results on OpenBLAS as on MKL + libsvml.so. Recompiling to MKL + libsvml.so will probably get numbers back in mid-sixties.

Things that help:
1) use Intel's OpenMP
2) use -march=native compiler option
3) OMP_PLACES=sockets and OMP_PROC_BIND=master,master,master seems to help.
4) MKL + libsvml.so (not tested in this comment -- see earlier comment above)

I am not understanding why parallelism doesn't help that much. If I want to investigate further I'll have to compile my own benchmarks in C to test out OpenMP capabilities on this computer.

**Comment by raver119** [comment]
o.n.n.NativeOps - Number of threads used for linear algebra 5  <-- this doesn't really matters, it won't be shown twice anymore, after i merge new branches.

About things that help:
1) Sure, Intel rocks.
2) Sure, native compilation rocks.
3) Surprisingly i can't confirm real OMP_PLACES real effect, and as for proc_bin, proc_bind(close) works better for both environments where i had tests. So for NativeOps we use close, and for Blas calls, env variable matters.
4) Sure, Intel rocks. But as i remember, libmvec produced pretty similar results for Paul

**Comment by raver119** [comment]
Also, can i see your **current** source code you're running there? I'm unable to confirm your double cpu use issue here. But suspects are: ParallelWrapper being used, G1/CMS gc being used.

**Comment by ChessPoker** [comment]
My [source](https://gist.github.com/ChessPoker/599f253941f4f3044bcfa3da5f4cdf87)
I'm running from r119_affinity branch from a couple of days ago. 
GC I assume is the default. I'm not specifying anything there.

**Comment by raver119** [comment]
Should be fixed in 0.6.0 and improved in 0.6.1-SNAPSHOT
